src/konserve/filestore.clj

```diff
@@ -251,46 +251,46 @@
                                :key       key
                                :exception e}))))))
 
-
-(defn- read-binary-old [f res-ch folder fn key locked-cb]
+(defn- read-binary-old [f folder fn key copy-fn]
+  (let [res-ch-old (chan)]
     (try
-          (let [ac (AsynchronousFileChannel/open (.getPath (FileSystems/getDefault)
-                                                           (str folder "/B_" fn)
-                                                           (into-array String []))
-                                                 (into-array StandardOpenOption
-                                                             [StandardOpenOption/READ]))
-                bb (ByteBuffer/allocate (.size ac))]
-            (.read ac
-                   bb
-                   0
-                   nil
-                   (proxy [CompletionHandler] []
-                     (completed [res att]
-                       (let [bais (ByteArrayInputStream. (.array bb))]
-                         (try
-                           (locked-cb {:input-stream bais
-                                       :size (.length f)
-                                       :file f})
-                           (catch Exception e
-                             (ex-info "Could not read key."
-                                      {:type :read-error
-                                       :key key
-                                       :exception e}))
-                           (finally
-                             (close! res-ch)
-                             (.close ac)))))
-                     (failed [t att]
-                       (put! res-ch (ex-info "Could not read key."
-                                             {:type :read-error
-                                              :key key
-                                              :exception t}))
-                       (close! res-ch)
-                       (.close ac)))))
-          (catch Exception e
-            (put! res-ch (ex-info "Could not read key."
-                               {:type :read-error
-                                :key key
-                                :exception e})))))
+      (let [ac (AsynchronousFileChannel/open (.getPath (FileSystems/getDefault)
+                                                       (str folder "/B_" fn)
+                                                       (into-array String []))
+                                             (into-array StandardOpenOption
+                                                         [StandardOpenOption/READ]))
+            bb (ByteBuffer/allocate (.size ac))]
+        (.read ac
+               bb
+               0
+               nil
+               (proxy [CompletionHandler] []
+                 (completed [res att]
+                   (let [bais (ByteArrayInputStream. (.array bb))]
+                     (try
+                       (copy-fn {:input-stream bais
```

Comment:
This is not a copy function per se, it is a function that will be called with read access at some point in a locked context. For this reason locked-cb is a more precise name in my opinion.